### ISSUE TITLE
Num instances for Min and Max?

### ISSUE THREAD

**Issue by jcpetruzza** [ekmett/semigroups#46]
`Sum` and `Prod` already have `Num` instances, which I find handy. It would be nice to have them for `Min` and `Max` as well.

**Comment by MarkZander** [comment]
I could use this as well. When attempting to use Min Float I get the following error:

    No instance for (Bounded Float) arising from a use of ‘foldMap’

I thought one reason for a semigroup was to include a Min and Max for non bounded numbers such as Integer. I was quite surprised to find out that Min and Max required bounded.

Float is bounded, but not listed as such for Haskell. Nothing that an orphaned instance wont' take care of.

**Comment by hvr** [comment]
This is needed to be able to define a `Monoid` instance:

```haskell
instance (Ord a, Bounded a) => Monoid (Min a) where
  mempty = maxBound
  mappend = (<>)
```

the problem is that `foldMap` requires a `Monoid`... so that's that

**Comment by MarkZander** [comment]
Understood that foldMap  needs a Monoid, but it would be nice if Min or Max could be written without Bounded.

The following do not have Bounded instances and yet are great candidates for Min / Max:

Float
Double
Integer
Arg

Arg is the interesting since it is defined in Data.Semigroup to have the following definitions:

type ArgMin <> a b = Min <http://hackage.haskell.org/package/semigroups-0.16.2.2/docs/Data-Semigroup.html#t:Min> (Arg <http://hackage.haskell.org/package/semigroups-0.16.2.2/docs/Data-Semigroup.html#t:Arg> a b) Source <http://hackage.haskell.org/package/semigroups-0.16.2.2/docs/src/Data-Semigroup.html#ArgMin>
type ArgMax <> a b = Max <http://hackage.haskell.org/package/semigroups-0.16.2.2/docs/Data-Semigroup.html#t:Max> (Arg <http://hackage.haskell.org/package/semigroups-0.16.2.2/docs/Data-Semigroup.html#t:Arg> a b) 

And yet it does not have Bounded defined either, meaning that ArgMax and ArgMin are somewhat useless as defined.

I thought part of the motivation behind semigroups and Option was that we could have Min and Max defined on Numbers that aren’t bounded.

Mark

**Comment by ekmett** [comment]
You can perform minimum and maximum using foldr1 or foldl1 without incurring `Bounded` constraints.

```haskell
foo = foldr1 (<>)
```

Using the `semigroupoids` package you can use Data.Semigroup.Foldable to

```haskell
getMin . foldMap1 Min
```

to fold over a container with at least one element and extract an answer.

On the other hand, we can use `Option (Min a)` to add a 'missing element' notion.

```haskell
fmap getMin . getOption . foldMap (Option . Just . Min)` :: (Foldable f, Ord a) => f a -> Maybe a
```

The point here is that `Min` and `Max` are about supplying the _Semigroup_ for minimum and maximum value. It is perfectly reasonable to extend this to a `Monoid` by trying to add a unit on top. This is what `Option` does here.

But because `Option` exists we don't need to bake it in and make everyone pay for it, even when they don't need it -- e.g. when they happen to be `Bounded`.

**Comment by ekmett** [comment]
**tl;dr** You are looking for `Option (Min a)` and `Option (Max a)` when you want me to magic up a unit.

**Comment by ekmett** [comment]
@jcpetruzza: Sure. I have no issue with adding instances for those. 

If someone put together a pull request adding "the common sense instances" that are missing, I'd be happy to accept it.

**Comment by ekmett** [comment]
Fixed by #51
